### PR TITLE
[SYSADMIN ACTION][Feature:InstructorUI] websockets in office hours queue

### PR DESCRIPTION
**Pull request by marwanatef2**
### What is the current behavior?
The office hours queue makes a request every 30 seconds to check if there is an update in the queue, and if so it renders the parts of the page that needs to get updated without refreshing the whole page.

### What is the new behavior?
Introducing websockets in the office hourse queue will only update the page on new socket messages without the need to ask (request) every 30 seconds whether there is an update or not.

### Other information
To enable websockets in order to test this PR please see [Docs](https://submitty.org/sysadmin/version_notes/v20.06.00)

[ draft docs PR : https://github.com/Submitty/submitty.github.io/pull/272 ]

### CODE REVIEW

File: .setup/install_system.sh [hunk 1]
```diff
@@ -517,7 +517,6 @@ EOF
     DISABLED_FUNCTIONS+="php_ini_loaded_file,readlink,symlink,link,set_file_buffer,proc_close,proc_terminate,"
     DISABLED_FUNCTIONS+="proc_get_status,proc_nice,getmyuid,getmygid,getmyinode,putenv,get_current_user,"
     DISABLED_FUNCTIONS+="magic_quotes_runtime,set_magic_quotes_runtime,import_request_variables,ini_alter,"
-    DISABLED_FUNCTIONS+="stream_socket_client,stream_socket_server,stream_socket_accept,stream_socket_pair,"
```
Comment (MasterOdin) [inline .setup/install_system.sh:520]: Let's remove the least amount of things from this list, starting with `stream_socket_client`, and not just the whole line (unless the whole line is necessary to have things work?).

File: site/app/libraries/socket/Server.php [hunk 2]
```diff
@@ -72,21 +77,19 @@ private function checkAuth(ConnectionInterface $conn): bool {
 
     /**
      * Push a given message to all or all-but-sender connections
+     * @param ConnectionInterface $from
+     * @param string $content
+     * @param string $course_name
+     * @return void
      */
-    private function broadcast(ConnectionInterface $from, string $content, string $course_name): void {
+    private function broadcast(ConnectionInterface $from, $content, $course_name): void {
```
Comment (MasterOdin) [inline site/app/libraries/socket/Server.php:85]: why remove the inline type hints?

Comment (marwanatef2) [reply site/app/libraries/socket/Server.php:85]: i added those type hints in the annotation right above the function, thought that was enough then.

Comment (MasterOdin) [reply site/app/libraries/socket/Server.php:85]: It's not, and we should be having inline types as much as possible, especially new classes / files, where we can do `declare(strict_types=1);` at the top, like here. If a file is declared strict, calling a function with the wrong argument types will cause an error to be raised if you call the function with the wrong types (and if you have the inline types but not strict typing, PHP will try and silently coerce types which I dislike).

See [this doc](https://www.php.net/manual/en/functions.arguments.php#functions.arguments.type-declaration.strict) for more info on strict typing.

Comment (marwanatef2) [reply site/app/libraries/socket/Server.php:85]: will add inline type declaration.

File: site/app/controllers/OfficeHoursQueueController.php [hunk 3]
```diff
@@ -459,4 +468,11 @@ public function showNewStatus() {
             )
         );
     }
+
+    private function sendSocketMessage($msg_array) {
+        $msg_array['user_id'] = $this->core->getUser()->getId();
+        $msg_array['course'] = $this->core->getConfig()->getCourse();
+        $client = new WebSocket\Client("ws://127.0.0.1:41983/");
+        $client->send(json_encode($msg_array));
+    }
```
Comment (Kevin-MM) [inline site/app/controllers/OfficeHoursQueueController.php:477]: Please add documentation and type hints for this function.

**Comment by marwanatef2** [comment]
This is because i updated `install_system.sh` , have you ran `submitty_install clean` after switching to this branch?

**Comment by MasterOdin** [comment]
Kevin's comment reminds me that we'll need to make two documentation changes for submitty.org:

1. Edit https://github.com/Submitty/submitty.github.io/blob/master/_docs/sysadmin/installation.md and update "Edit PHP Settings" and remove `stream_socket_client` from the list of disabled functions
2. Add a new page to https://github.com/Submitty/submitty.github.io/tree/master/_docs/sysadmin/version_notes that documents the above change to php.ini file

For (2), the file name would be the expected release this will land in (20.07.00) and then a short bit on what a sysadmin would need to manually do to enable this feature.

**Comment by bmcutler** [comment]
@marwanatef2 and @elihschiff 
As expected, lots of merge conflicts with master.  Please resolve, retest, and reapprove :)

**Comment by elihschiff** [comment]
I am trying to test this pr but I keep getting a websocket error. I am not sure if the issue is from something on my end or if the pr is broken (it is failing the tests atm not sure if that is related)
When I try and join the queue I get this error 
image

I also made sure to follow all the steps in [this doc page](https://submitty.org/sysadmin/version_notes/v20.06.00)

One thing that may or may not be related is when I run `systemctl status submitty_websocket_server` my output looks like this. It says it is active however it seems strange that it is outputting that code at the bottom, I have never seen that before and I cannot figure out why exactly it is happening

image

**Comment by bmcutler** [comment]
@elihschiff Did you also follow the draft sysadmin version notes for this PR linked in the top post?

**Comment by elihschiff** [comment]
@bmcutler yes I tried that, everything was already correct but when I tested, but I redid all the steps just in case and it did not fix the issue
If it is working for people other than me I can try running a new fresh vagrant up and see if that fixes it.

**Comment by elihschiff** [comment]
oh wait I missed it, there are 2 version of the docs. My bad let me try the draft version, I was confused

File: tests/e2e/test_office_hours_queue.py [hunk 4]
```diff
@@ -288,7 +293,7 @@ def expectedAlerts(self, success=0, error=0, success_text=[], error_text=[]):
         self.assertEqual(self.countAlertError(error_text), error)
 
 def enableQueue(self):
-    self.log_in(user_id='instructor')
+    #self.log_in(user_id='instructor')
```
Comment (elihschiff) [inline tests/e2e/test_office_hours_queue.py:296]: I think this line should just be removed, I dont see a reason to keep it in the code but commented out.

Comment (marwanatef2) [reply tests/e2e/test_office_hours_queue.py:296]: oh yeah, i just forgot to remove it

File: site/app/libraries/database/DatabaseQueries.php [hunk 5]
```diff
@@ -5793,10 +5793,6 @@ public function changeQueueToken($token, $queue_code) {
         $this->course_db->query("UPDATE queue_settings SET token = ? WHERE code = ?", [$token, $queue_code]);
     }
 
-    public function getLastQueueUpdate() {
```
Comment (MasterOdin) [inline site/app/libraries/database/DatabaseQueries.php:5796]: Travis is failing as this function is still used in one other place. Should we be keeping this function then, or refactor the one place using it?

> ```
> ------ --------------------------------------------------------------- 
>  Line   models/OfficeHoursQueueModel.php                               
> ------ --------------------------------------------------------------- 
>  181    Call to an undefined method                                    
>         app\libraries\database\DatabaseQueries::getLastQueueUpdate().  
> ------ --------------------------------------------------------------- 
>```

**Comment by Kevin-MM** [comment]
Hi @marwanatef2 , can you please correct the failing lint and other tests.

File: site/app/libraries/database/DatabaseQueries.php (hunk 5 above)
Comment (marwanatef2) [reply site/app/libraries/database/DatabaseQueries.php:5796]: this function is not called any more, i just forgot to remove it

**Comment by marwanatef2** [comment]
> can you please correct the failing lint and other tests.

Done.

File: site/app/libraries/socket/Server.php [hunk 6]
```diff
@@ -17,32 +17,37 @@ class Server implements MessageComponentInterface {
...
      */
     private function checkAuth(ConnectionInterface $conn): bool {
         $request = $conn->httpRequest;
-        $client_id = $conn->resourceId;
-        $origin = $request->getHeader('origin')[0];
+        $user_agent = $request->getHeader('User-Agent')[0];
+
+        if ($user_agent === 'websocket-client-php') {
+            return true;
+        }
```
Comment (Kevin-MM) [inline site/app/libraries/socket/Server.php:50]: Are there any security implications of this?  Would it be possible for a bad actor to fake their user agent to be `websocket-client-php` and gain authorization to the websocket server?

File: site/app/templates/officeHoursQueue/AnnouncementMsg.twig [hunk 7]
```diff
@@ -0,0 +1,7 @@
+{% if viewer.getQueueAnnouncementMessage() | length > 0 %}
+    <div id="announcement">
+        <h2>Office Hours Queue Announcements:</h2>
+        <p>{{viewer.getQueueAnnouncementMessage() | markdown}}</p>
+    </div>
+    <br>
```
Comment (Kevin-MM) [inline site/app/templates/officeHoursQueue/AnnouncementMsg.twig:6]: Please remove the line break.

File: site/app/controllers/OfficeHoursQueueController.php [hunk 8]
```diff
@@ -481,8 +481,38 @@ public function updateAnnouncement() {
...
+            )
+        );
+    }
+
+    private function sendSocketMessage($msg_array) {
+        $msg_array['user_id'] = $this->core->getUser()->getId();
+        $msg_array['page'] = $this->core->getConfig()->getCourse() . "-office_hours_queue";
+        $client = new WebSocket\Client("ws://127.0.0.1:41983/");
+        $client->send(json_encode($msg_array));
+    }
```
[the remark this answers is not in the document]
Comment (MasterOdin) [reply site/app/controllers/OfficeHoursQueueController.php:517]: While I agree with this, within this PR, let's just abstract this into making a new class under `site/app/libraries/socket/Client.php` which is (roughly):

```php
<?php

declare(strict_types=1);

namespace app\libraries\socket;

class Client extends \WebSocket\Client {
    public function __construct() {
        parent::__construct('ws://127.0.0.1:41983/');
    }
}
```

for now. I'll write-up an issue later on today (or this week) that details configuration settings for the client, and that can be strictly dealt with in an independent PR.

File: site/app/libraries/socket/Server.php (hunk 6 above)
Comment (MasterOdin) [reply site/app/libraries/socket/Server.php:50]: Yes. This has security implications. Using the above code, we should validate that if the user_agent is `websocket-client-php` and that `$request->getHeader('Session-Secret')[0] === $this->core->getConfig()->getSecretSession()`.

**Comment by marwanatef2** [comment]
Thank you for your reviews @Kevin-MM @MasterOdin , I'll address those requested changes once we fix this travis issue.

File: site/app/libraries/socket/Server.php [hunk 9]
```diff
@@ -17,32 +17,40 @@ class Server implements MessageComponentInterface {
...
+
+        if ($user_agent === 'websocket-client-php') {
+            $session_secret = $request->getHeader('Session-Secret')[0];
+            if ($session_secret  === $this->core->getConfig()->getSecretSession()) {
+                return true;
+            }
+        }
 
         $cookieString = $request->getHeader("cookie")[0];
         parse_str(strtr($cookieString, ['&' => '%26', '+' => '%2B', ';' => '&']), $cookies);
```
Comment (Kevin-MM) [inline site/app/libraries/socket/Server.php:56]: Would it be appropriate to add an `else { return false; }` here?  I imagine the php-client doesn't have a cookie set and I'm just wondering if its possible we might end up trying to run functions on null values and result in an exception.

Comment (marwanatef2) [reply site/app/libraries/socket/Server.php:56]: yeah, makes sense

**Comment by marwanatef2** [comment]
Thanks for this detailed feedback/test/code review @Kevin-MM , please retest and tell me if anything needs further changes, I hope we're good to go now.

**Comment by marwanatef2** [comment]
This is exactly the error i fixed in the last commit, Are you sure you pulled the last changes?
If so can you show me the **form action** for joining the queue just before you click join (when you know it will give you the _invalid secret code_ error message) @Kevin-MM

**Comment by bmcutler** [comment]
@MasterOdin 
Do you have further requested changes on this sockets office hours PR?  
Or can you approve and we get it merged?

**Comment by bmcutler** [comment]
@MasterOdin or @marwanatef2 
Please fix the eslint error.

**Pull request merged**